controllers/taskcluster_operations.go

```diff
@@ -936,3 +942,41 @@ func (o *TaskClusterOperations) ensureCrypto(name string) {
 func (o *TaskClusterOperations) getCrypto(name string) CryptoConfig {
 	return o.ensureServiceAccount(name).CryptoConfig
 }
+
+func (o *TaskClusterOperations) FinishDeployment(ctx context.Context) (reconcile.Result, error) {
+	upgradeKey := types.NamespacedName{
+		Namespace: o.dbUpgradeJob.Namespace,
+		Name:      o.dbUpgradeJob.Name,
+	}
+
+	var job batchv1.Job
+
+	err := o.Client.Get(ctx, upgradeKey, &job)
+	if err == nil {
+		// Check if it has finished.
+		desiredCompletions := int32(1)
+		if job.Spec.Completions != nil {
+			desiredCompletions = *job.Spec.Completions
+		}
+
+		completions := job.Status.Failed + job.Status.Succeeded
+		if completions < desiredCompletions {
+			o.Logger.Info("waiting for migration job to complete")
+			return reconcile.Result{
+				RequeueAfter: time.Minute,
+			}, nil
+		}
+
+		// Delete the old Job if it doesn't match.
+		if job.Annotations == nil || job.Annotations[hashAnnotation] != o.dbUpgradeHash {
+			o.Logger.Info("deleting old migration job")
+			if err := o.Client.Delete(ctx, &job); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	} else if !apierrors.IsNotFound(err) {
+		return reconcile.Result{}, err
+	}
```

Comment:
is there a reason why this can't "early out", allowing the happy path to be dedented?

Reply:
I did it this way in case we ever needed to add any extra logic here. In this case, the path right before the final return is a straight-forward place to add new stuff. I might swap it for now and we can just break it into functions if needed.